File: MinStack/min_stack.py
class MinStack:
    def __init__(self):
        _stack = []
        _min_stack = []

    def push(self, val: int) -> None:
        self._stack.append(val)
        m = min(self._min_stack[-1], val) if self._min_stack else val
        self._min_stack.append(m)

    def pop(self) -> None:
        self._stack.pop()
        self._min_stack.pop()

    def top(self) -> int:
        return self._stack[-1]

    def get_min(self) -> int:
        return self._min_stack[-1]

File: MinStack/MinStack.hs
data MinStack = MinStack {stack :: [Int], _minStack :: [Int]} deriving (Show)

empty :: MinStack
empty = MinStack [] []

push :: Int -> MinStack -> MinStack
push i (MinStack [] []) = MinStack [i] [i]
push i (MinStack (s : ss) (m : ms)) =
  MinStack (i : s : ss) (min i m : m : ms)

pop :: MinStack -> MinStack
pop (MinStack [] []) = MinStack [] []
pop (MinStack (s : ss) (m : ms)) = MinStack ss ms

top :: MinStack -> Int
top (MinStack [] []) = error "Empty stack"
top (MinStack (s : _) _) = s

getMin :: MinStack -> Int
getMin (MinStack [] []) = error "Empty stack"
getMin (MinStack _ (m : _)) = m
